### ISSUE TITLE
Invisible `\0` character in REPL if `Ctrl-D` is pressed when there isn't a prompt

### ISSUE THREAD

**Issue by yuyichao** [JuliaLang/julia#11027]
See following REPL log with comment

``` julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+4512 (2015-04-26 10:59 UTC)
 _/ |\__'_|_|_|\__'_|  |  Commit 98e87ab* (1 day old master)
|__/                   |  x86_64-unknown-linux-gnu

# Press Ctrl-D
julia> # Pressing Ctrl-D here does not quit the REPL

julia> sleep(10)

# Press Ctrl-D
julia> ? # Pressing ? here does not bring up the help prompt

julia> sleep(10)

# Press Ctrl-D
help?> sleep # Ctrl-B/Ctrl-A and type ?sleep
ERROR: ArgumentError: Symbol name may not contain \0
 in anonymous at ./REPL.jl:685
 in anonymous at ./REPL.jl:595
# Error popup and killed the REPL, error messages suggested that there's a invisible "\0" character in the string.
```

I've also seen once when `sleep` doesn't work anymore (not waiting long enough) but haven't been able to reproduce it yet.

**Comment by yuyichao** [comment]
Actually just reproduces the `sleep` issue. So having a `\0` in REPL breaks other stuffs as well and they might be present in the history and hard to get rid of....

```julia
julia> sleep(2) # Normal prompt

# Press Ctrl-D
julia> sleep(3) # type "sleep(3)", actual string read "\0sleep(3)"

# Does not sleep
julia> sleep(3) # Use Ctrl-P to navigate to the previous command, actually string read "\0sleep(3)"

# Does not sleep (So the \0 is actually saved in history)
julia> sleep(3) # Type "sleep(3)"

# Working
julia> sleep(3) # Type "sleep(3)"

# Working
# Type Ctrl-D
julia> @time sleep(3) # Type ...

# No output
julia> @time sleep(3) # Ctrl-P

# No output
julia> @time sleep(3) # Ctrl-P and edit it tobe "@time \0sleep(3)", this can happen when editing "\0sleep(3)" in the history
ERROR: wrong number of arguments
```

**Comment by yuyichao** [comment]
And here is part of my `.julia_history`

```julia
# time: 2015-04-27 08:37:34 EDT
# mode: julia
	^@@time sleep(3)
# time: 2015-04-27 08:37:52 EDT
# mode: julia
	@time ^@^@sleep(3)
```

where `^@` represent a invalid character in Emacs. I sometimes press `Ctrl-D` more than once during the test hence the `^@^@`.

**Comment by blakejohnson** [comment]
Related to #7001.

**Comment by thepulkitagarwal** [comment]
Also related to #7605

**Comment by rfourquet** [comment]
That seems fixed? (maybe by #21505?)

**Comment by yuyichao** [comment]
Seems like it.
